### ISSUE TITLE
Explicit annotation as type[Any] is inconsistently ignored in favour of inferred type 

### ISSUE THREAD

**Issue by FasterSpeeding** [microsoft/pyright#2740]
**Describe the bug**
The inferred type inconsistently takes priority over explicit annotation as `type[Any]` when resolving the type for typed assignments.

**To Reproduce**
* Run the example code through the latest Pyright release
* Watch it give the following results
```
Found 1 source file
C:\Users\Lucina\PycharmProjects\tanjun\test.py
  C:\Users\Lucina\PycharmProjects\tanjun\test.py:6:13 - info: Type of "foo" is "Mock"
  C:\Users\Lucina\PycharmProjects\tanjun\test.py:8:13 - info: Type of "bar" is "Type[int]"
  C:\Users\Lucina\PycharmProjects\tanjun\test.py:11:13 - info: Type of "baz" is "Any"
0 errors, 0 warnings, 3 infos
```

**Expected behavior**
The behaviour should be consistent with how the example baz variable is treated where the type hint of `Any` takes priority over the inferred type of `Literal[42]` meaning that both foo and bar should be revealed to be `type[Any]`.

**Screenshots or Code**
If applicable, add screenshots or the text of the code (surrounded by triple back ticks) to help explain your problem.
```python
import typing
import unittest.mock


foo: type[typing.Any] = unittest.mock.Mock()
reveal_type(foo)
bar: type[typing.Any] = int
reveal_type(bar)

baz: typing.Any = 42
reveal_type(baz)
```

**VS Code extension or command-line**
Pyright 1.1.197 as a command-line tool.
Targeting python 3.10.

**Additional context**
This behaviour was introduced by v1.1.194

**Comment by erictraut** [comment]
What you're observing here is type narrowing that is applied on assignment. For example:
```python
foo: object
reveal_type(foo) # object
foo = 3
reveal_type(foo) # Literal[3]
```

A symbol with a declared type of `Any` is not narrowed.

In the case of `type[Any]`, it's not clear whether the normal rules for type narrowing or the "don't narrow `Any`" exception applies. The type `type[Any]` is equivalent to `type`. The current implementation treats it like `type` in this case, and it narrows accordingly.

Mypy is somewhat inconsistent in how it handles `type` and `type[Any]`, so I'm not sure we can look to mypy for an answer here.

We have three options:
1. Treat `type[Any]` as `type` and do apply type narrowing as though it is `type` in both cases.
2. Treat `type[Any]` as `type` and do not apply type narrowing in either case.
3. Treat `type[Any]` as different from `type` and apply type narrowing only in the latter case.

Currently pyright uses option 1.

I think we can cross off option 2 because it will create a lot of problems for existing code, and I think that behavior would be surprising to many users.

You're arguing for option 3.

I'll need to think about this a bit more. My current inclination is to stick with option 1, if for no other reason than changing to option 3 will probably cause new errors for some existing pyright users.

**Comment by erictraut** [comment]
After thinking about this more, I think option 1 is the correct answer. That's the current behavior in pyright. If you want to make an argument for switching to option 3, I'm willing to reconsider. For now, I'm going to mark this as closed.

**Comment by FasterSpeeding** [comment]
I can understand if this isn't exactly an acceptable use case but for what it's worth Pyright 1.1.197 introduced new errors for code I had in some of my projects where I was using a `unittest.mock.Mock()` as a fake type in a similar fashion as to what's shown below

```py
import typing
import unittest.mock

mock_type: type[typing.Any] = unittest.mock.Mock()
reveal_type(mock_type)
```

but since that release mock_type would now be resolved with `unittest.mock.Mock` without any error or acknowledgment of that type hint which is at the very least kinda weird behaviour and doesn't match MyPy (0.930) which seems to leave mock_type as `type[typing.Any]` for the same scenario

Sorry for the late response btw.

**Comment by erictraut** [comment]
It sounds like you want `mock_type` to have a type of `Any` rather than `type[Any]`.
